### PR TITLE
IMP: Add ability to adjust heatmap colorscale

### LINKED ISSUE #148: confusion-matrix: add parameter for adjusting color scale?
**Improvement Description**
Add parameter (or automatically) adjust scale, e.g., to set colorscale range from 0 to 1

**Current Behavior**
Colors scale to accuracy range, which can be misleading if accuracy rates are consistent across classes.

**Proposed Behavior**
Add a parameter to allow colorscale adjustment, and/or automatically scale colorscale

### PR DESCRIPTION
**Pull request by Oddant1**
Closes #148 Adds ability to adjust colorscale of heatmaps derived from confusion matrices via seaborn's vmin and vmax parameters. Is this along the lines of what you want @nbokulich? Once the parameters are properly implemented I can add some tests.

### CODE REVIEW

File: q2_sample_classifier/plugin_setup.py [hunk 1]
```diff
@@ -454,6 +456,8 @@
     parameter_descriptions={
         'truth': 'Metadata column (true values) to plot on y axis.',
         'missing_samples': parameter_descriptions['base']['missing_samples'],
+        'vmin': 'The minimum color value for the heatmap',
```
Comment (nbokulich) [inline q2_sample_classifier/plugin_setup.py]: How about we modify from the seaborn descriptions:
'vmin': 'The minimum value to use for anchoring the colormap. If "auto", vmin is set to the minimum value in the data.', 
'vmax': 'The maximum value to use for anchoring the colormap. If "auto", vmax is set to the maximum value in the data.'

**Comment by thermokarst** [comment]
@nbokulich - the first thing that comes to my mind RE the unintentional masking of data is to promote that to a real mask (using the `mask` parameter in `sns.heatmap`).

**Comment by nbokulich** [comment]
> promote that to a real mask (using the mask parameter in sns.heatmap).

I suppose that works. @Oddant1 what do you think about adding a mask to `_plot_heatmap_from_confusion_matrix` if vmax or vmin is not None? See `https://seaborn.pydata.org/generated/seaborn.heatmap.html` for more details on masking.

**Comment by Oddant1** [comment]
@nbokulich and @thermokarst, if I'm reading this correctly, what's happening is it's too easy to set `vmin` and `vmax` to values that cause some of the cells to fade into the background unintentionally. If this is the case, I'm not sure I understand how the mask is going to solve this. How would adding a mask change the appearance of this heatmap? 
image
It says the mask allows you to intentionally cover up cells, but I don't see how allowing you to do it intentionally stops it from happening unintentionally. I apologize for my lack of understanding of how this particular utility works.

**Comment by nbokulich** [comment]
@Oddant1 thank you for asking — I scratched my head for a bit first at @thermokarst 's suggestion. There are really two problems here: vmin/vmax can lead to unintentional (or nefarious) "masking" (in my parlance) of data that are off-scale, causing two distinct issues:
1. off-scale values cannot be distinguished so datapoints can look higher or lower than they actually are.
2. values that really are zero will appear to be vmin. E.g., in the heatmap above it looks like most squares are 0.9 when most are 0.0.

masking (in seaborn parlance, not mine) will allow values lower than vmin to be dropped, so that the square appears white instead of whatever the lowest color value is. This is effectively a fix for issue 2, but unfortunately not for 1. However, issue 1 might not be our job to solve just as it's not seaborn's job; e.g., users should be responsible and pay attention, it is a mostly foolproof step, and we can't run too much of a nanny framework here. Even if a user accidentally uses an inappropriate vmin value (e.g., they are copying/pasting from an existing workflow), the masked values will make it apparent to them that there's a problem and they need to re-run that step.

So this is a "good enough" fix in my mind, even if it's not perfect.

**Comment by Oddant1** [comment]
@nbokulich so it wouldn't eliminate the issue, but it would make it obvious that the issue had occurred which would hopefully get the user to rerun and pay closer attention to their `vmin` and `vmax` values

**Comment by nbokulich** [comment]
yep! I am also on the fence about whether we should expose a `mask` parameter to toggle this behavior. I'm thinking no for now and if it bugs us or users, we can add at a later date. What do you and @thermokarst think?

**Comment by thermokarst** [comment]
One way to expose it in a limited manner is to use `TypeMap`!

```python
A, B, __ = TypeMap({
    (Float, Bool % Choices(False)): Visualization,
    (Str % Choices('auto'), Bool): Visualization,
})

...

parameters={
    'foo': A,
    'bar': B
...

```

This produces  a signature like this:

image

Let me know if you want the example expanded on.

**Comment by Oddant1** [comment]
So in that example, `vmin` and `vmax` look like `foo`, and `mask` looks like `bar`?

**Comment by Oddant1** [comment]
https://github.com/qiime2/q2-sample-classifier/pull/170/commits/977b1bbce070cad64976fbdccda9f452c92d2ef8 Not sure why the tests in this commit fail (the regex's aren't matching for some reason, I'm sure I'm just missing something stupid), but is this this along the lines of what we want?

**Comment by nbokulich** [comment]
yep! that looks like what we need — but pull out the vmin/vmax check into a separate utility function and call it as early as possible in the `_plot_confusion_matrix` function, rather than in the `_plot_heatmap_from_confusion_matrix` function.

File: q2_sample_classifier/plugin_setup.py (hunk 1 above)
Comment (nbokulich) [reply q2_sample_classifier/plugin_setup.py]: @Oddant1 please see this old change needed

File: q2_sample_classifier/visuals.py [hunk 2]
```diff
@@ -149,6 +153,27 @@ def _plot_confusion_matrix(y_test, y_pred, classes, normalize, palette):
...
 
 
+def _check_vmin_and_vmax(cm, vmin, vmax):
+    lowest_frequency = np.amin(cm)
+    highest_frequency = np.amax(cm)
+
+    error = ''
+    if vmin is not None:
+        if vmin > lowest_frequency:
+            error += ('Your vmin value must be less than or equal to the '
```
Comment (nbokulich) [inline q2_sample_classifier/visuals.py]: please change to:
"vmin must be less than or equal to the lowest predicted class frequency"

File: q2_sample_classifier/visuals.py [hunk 3]
```diff
@@ -149,6 +153,27 @@ def _plot_confusion_matrix(y_test, y_pred, classes, normalize, palette):
...
+                      'lowest prediction frequency present:\n'
+                      f'\t{vmin!r} is greater than {lowest_frequency!r}')
+    if vmax is not None:
+        if vmax < highest_frequency:
+            if error:
+                error += '\n'
+            error += ('Your vmax value must be greater than or equal to the '
```
Comment (nbokulich) [inline q2_sample_classifier/visuals.py]: Please change to:
"vmax must be greater than or equal to the highest predicted class frequency"

**Pull request merged**